src/docs/process-binding.md

```diff
@@ -0,0 +1,87 @@
+- Marhshaling handled differently based on the type of the parameter and the type of the function
+- Implicit arguments
+- Versioning based on the `@available` attribute.
+
+For this reason I strongly recommend using code-generation tools that can work in a non-linear fashion. There are several ways to achieve this, but I would strongly recommend using the Dynamo framework from Binding Tools for Swift as it can handle both C# and Swift and generates non-linearly. In addition, it shouldn't be a stretch to generate code in parallel on type boundaries.
+
+Because of this, I think we should adopt a strategy and factory pattern for handlers at various levels.
+
+The general pattern would work like this:
+1. Start with a Swift language entity
+2. Aggregate information about that entity
+3. Select a factory to create a handler for that entity
+4. The handler will generate a context object for handling that entity
+5. Execute a series of steps through the handler that will do work apropriate for each step.
+6. Aggregate the result and generate as needed.
+
+Handlers will contain factories for handling sub steps, if needed.
+
+For example, given the following Swift code:
+```Swift
+public func generateAClass(String name) -> SomeClass { }
+```
+The process would look at this and identify this as a top-level function and will select a handler factory for it.
+The handler will create a context for the object which would include a class for the top-level object to live in (C# doesn't have top-level functions) and a class to hold top-level pinvokes and a function generation context which would include a place to place function argument declarations, generic declarations, function argument pre-marshaling code, pinvoke argument declarations, pinvoke argument expressions, post-marshaling code, return type declaration, and a return expression. 
+
+The handler will execute a step to name the function and the associated pinvoke, including the entry point and library.
+Then for each argument, it will gather information about each argument and from the function handler get a factory to build an argument handler for type `String`. This will in turn name the argument, generate the C# type and add it to the C# argument declaration. It will define the argument type for the pinvoke and add it to the C# pinvoke argument list. If needed, it will generate premarshal code and add it to the premarshal list and post marshal code, and finally an expression for calling the pinvoke.
+
+A similar process will be done for handling the return type and value. In this case, the pinvoke return type will be a `NativeHandle` and it will be used in conjunction with a registry to either retrieve an already existing C# object that is bound to that handle or it will build one through a factory.
+
+After all this is done, the function handler will finish up by aggregating all the information, writing the C# method and writing the C# pinvoke.
```

Comment:
In this example we do not need to generate Swift code, but what about cases where we will need to do this. Will it happen here as well?

Reply:
It makes sense to do so. The reason being that in any cases where marshaling is not 1:1 with C# capabilities, we will need to change the way that parameters are handled.  I'll expound on this more in the docs because it will make it clear why non-linear code writing will make tasks much easier.